src/transformStateWithClones.js

```diff
@@ -7,7 +7,58 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const history = [];
+
+  for (const obj of actions) {
+    switch (obj.type) {
+      case 'addProperties':
+        addProperties(stateCopy, obj.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(stateCopy, obj.keysToRemove);
+        break;
+      case 'clear':
+        clearObject(stateCopy);
+        break;
+    }
+  }
+
+  function addProperties(object, extraData) {
+    const copy = { ...object };
+
+    Object.assign(copy, extraData);
+
+    stateCopy = copy;
+
+    history.push(stateCopy);
+  }
+
+  function removeProperties(object, keysToRemove) {
+    const copy = { ...object };
+
+    for (const key of keysToRemove) {
+      delete copy[key];
+    }
+
+    stateCopy = copy;
+
+    history.push(stateCopy);
+  }
+
+  function clearObject(object) {
+    const copy = { ...object };
+
+    for (const key in object) {
+      delete copy[key];
+    }
+
+    stateCopy = copy;
```

Comment:
Reassigning `stateCopy` directly within the helper functions can lead to unintended side effects. Consider returning the modified `copy` from the `clearObject` function and updating `stateCopy` in the main function.